adal/src/main/java/com/microsoft/aad/adal/AuthenticationException.java

```diff
@@ -70,6 +84,89 @@ public AuthenticationException(ADALError code, String details) {
     public AuthenticationException(ADALError code, String details, Throwable throwable) {
         super(details, throwable);
         mCode = code;
+
+        if (null == throwable) {
+            return;
+        }
+
+        if (throwable instanceof AuthenticationException) {
+            mServiceStatusCode = ((AuthenticationException) throwable).getServiceStatusCode();
+            mResponseBody = ((AuthenticationException) throwable).getResponseBody();
+            mHttpResponseHeaders = ((AuthenticationException) throwable).getHttpResponseHeaders();
+        }
+    }
+
+    /**
+     * @param code Resource file related error code. Message will be derived
+     *            from resource with using app context
+     * @param details Details related to the error such as query string, request info.
+     * @param response HTTP web response
+     */
+    public AuthenticationException(ADALError code, String details, HttpWebResponse response) {
+        super(details);
+        mCode = code;
+
+        if (null == response) {
+            return;
+        }
+
+        mServiceStatusCode = response.getStatusCode();
+
+        if (null != response.getResponseHeaders()) {
+            mHttpResponseHeaders = response.getResponseHeaders();
```

Comment:
HashMapExtensions.getJsonResponse

Reply:
HashMapExtensions.getJsonResponse is used to convert the response body, which is a String, into Map<String, String>.